src/OpenSpaceToolkit/Astrodynamics/Trajectory/Orbit.cpp

```diff
@@ -185,192 +185,217 @@ Pass Orbit::getPassWithRevolutionNumber(const Integer& aRevolutionNumber, const
 
-                if (currentPass.isDefined())
+            if ((previousStateCoordinates_ECI_z > 0.0) && (currentStateCoordinates_ECI_z <= 0.0))
+            {
+                const Instant crossingInstant = Orbit::GetCrossingInstant(epoch, previousInstant, currentInstant, getZ);
+
+                if (isForwardPropagated)
                 {
-                    currentPass = {
-                        currentRevolutionNumber + 1,
-                        currentPass.accessInstantAtPassBreak(),
-                        northPointCrossing,
-                        descendingNodeCrossing,
-                        southPointCrossing,
-                        previousInstant,
-                    };
+                    descendingNodeCrossing = crossingInstant;
                 }
                 else
                 {
-                    const Instant ascendingNodeCrossing =
-                        Real(this->modelPtr_->calculateStateAt(this->modelPtr_->getEpoch())
-                                 .getPosition()
-                                 .accessCoordinates()
-                                 .z())
-                                .isNear(0.0, epsilon)
-                            ? this->modelPtr_->getEpoch()
-                            : Instant::Undefined();
-
-                    currentPass = {
-                        currentRevolutionNumber,
-                        ascendingNodeCrossing,
-                        northPointCrossing,
-                        descendingNodeCrossing,
-                        southPointCrossing,
-                        previousInstant,
-                    };
+                    previousInstant = crossingInstant;
+                    break;
                 }
             }
-            else  // Reverse propagation
-            {
-                throw ostk::core::error::runtime::ToBeImplemented(
-                    "Orbit::getPassWithRevolutionNumber > Reverse propagation"
-                );
-            }
 
-            if (currentPass.isDefined())
+            if ((previousStateCoordinates_ECI_z < 0.0) && (currentStateCoordinates_ECI_z >= 0.0))
             {
-                this->passMap_.insert({currentPass.getRevolutionNumber(), currentPass});
+                const Instant crossingInstant = Orbit::GetCrossingInstant(epoch, previousInstant, currentInstant, getZ);
+                if (isForwardPropagated)
+                {
+                    previousInstant = crossingInstant;
+                    break;
+                }
+                else
+                {
+                    descendingNodeCrossing = crossingInstant;
+                }
             }
+
+            previousStateCoordinates_ECI_z = currentStateCoordinates_ECI_z;
+            previousStateCoordinates_ECI_zdot = currentStateCoordinates_ECI_zdot;
+            previousInstant = currentInstant;
         }
 
-        if (currentPass.getRevolutionNumber() == aRevolutionNumber)
+        if (currentPass.isDefined())
```

Comment:
Not asking for a change in this MR, but we might be able to simplify a bit by separating out the initial case into a separate check at the beginning:
```
if _passMap.is_empty() {
   // Do all of the flow that is necessary when currentPass is not defined
}
```

Which may let us separate that logic into a separate function. Additionally, if we separate the:
```
while(true) {
    // Fill out incomplete passes
}
```
into it's own function, could be re-used by both flows.

Again, not asking for this. Just tossing ideas around to make this algorithm a little more transparent.